rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/ChargingRSpace.scala

```diff
@@ -80,52 +91,59 @@ object ChargingRSpace {
         Option[(ContResult[Par, BindPattern, TaggedContinuation], Seq[Result[ListParWithRandom]])]
       ] =
         for {
-          _       <- charge[F](storageCostProduce(channel, data))
+          _       <- charge[F](storageCostProduce(channel, data).copy(operation = "produces storage"))
           prodRes <- space.produce(channel, data, persist, sequenceNumber)
-          _       <- handleResult(prodRes)
+          _       <- handleResult(prodRes, Produce(data.randomState, persist))
         } yield prodRes
 
       private def handleResult(
           result: Option[
             (ContResult[Par, BindPattern, TaggedContinuation], Seq[Result[ListParWithRandom]])
-          ]
+          ],
+          triggeredBy: TriggeredBy
       ): F[Unit] =
         result match {
 
-          case None => Sync[F].unit
+          case None => charge[F](eventStorageCost(triggeredBy.channelsCount))
 
           case Some((cont, dataList)) =>
-            val refundForConsume =
-              if (cont.persistent) Cost(0)
-              else
-                storageCostConsume(cont.channels, cont.patterns, cont.value)
-
-            val refundForProduces = refundForRemovingProduces(
-              dataList,
-              cont.channels
-            )
-
-            val refundValue = refundForConsume + refundForProduces
-
             for {
-              _ <- if (refundValue == Cost(0))
-                    Sync[F].unit
-                  else charge[F](Cost(-refundValue.value, "storage refund"))
+              id <- consumeId(cont.continuation)
+
+              // We refund for non-persistent continuations, and for the persistent continuation triggering the comm.
+              // That persistent continuation is going to be charged for (without refund) once it has no matches in TS.
+              refundForConsume = if (!cont.persistent || id == triggeredBy.id) {
+                storageCostConsume(cont.channels, cont.patterns, cont.continuation)
+              } else {
+                Cost(0)
+              }
+              refundForProduces = refundForRemovingProduces(dataList, cont.channels, triggeredBy)
+
+              _ <- charge[F](Cost(-refundForConsume.value, "consume storage refund"))
+              _ <- charge[F](Cost(-refundForProduces.value, "produces storage refund"))
+              _ <- charge[F](eventStorageCost(triggeredBy.channelsCount))
+                    .unlessA(triggeredBy.persistent)
+              _ <- charge[F](commEventStorageCost(cont.channels.size))
             } yield ()
         }
 
       private def refundForRemovingProduces(
           dataList: Seq[Result[ListParWithRandom]],
-          channels: Seq[Par]
-      ): Cost =
-        dataList
+          channels: Seq[Par],
+          triggeredBy: TriggeredBy
+      ): Cost = {
+        val removedData = dataList
```

Comment:
Refund for produces should be 0 when it's triggered by a peek.
Deterministic peek cost can't be achieved and tested though unit the bug I've mentioned in the other comment is fixed (due to event storage costs being wrong).